### PR TITLE
Minor CMake change to better integrate c++17 and modernize GTest-integration

### PR DESCRIPTION
**Pull request by emmenlau**
This is a minor PR that improves the integration of c++17 and of testing, and specifically of GTest. It brings better compatibility with FindGTest from (future) cmake.

Please let me know if it would need more motivation? :-)

### CODE REVIEW

**Comment by emmenlau** [comment]
The AppVeyor build error seems unrelated, unless I'm mistaken? It says:
```
LINK: command "C:\PROGRA~2\MI0E91~1.0\VC\bin\amd64\link.exe /nologo @CMakeFiles\test_xtensor_blas.dir\objects1.rsp /out:test_xtensor_blas.exe /implib:test_xtensor_blas.lib /pdb:C:\projects\xtensor-blas\test\test_xtensor_blas.pdb /version:0.0 /MANIFEST:NO /debug /INCREMENTAL /subsystem:console C:\xtensor-conda\Library\mingw-w64\lib\libopenblas.dll.a C:\xtensor-conda\Library\lib\gtest_main-mdd.lib C:\xtensor-conda\Library\lib\gtest-mdd.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib" failed (exit code 1120) with the following output:
gtest-mdd.lib(gtest-all.cc.obj) : warning LNK4099: PDB 'gtestd.pdb' was not found with 'gtest-mdd.lib(gtest-all.cc.obj)' or at 'C:\projects\xtensor-blas\test\gtestd.pdb'; linking object as if no debug info
gtest-mdd.lib(gtest-all.cc.obj) : error LNK2019: unresolved external symbol __std_reverse_trivially_swappable_4 referenced in function "void __cdecl std::_Reverse_unchecked1<enum testing::internal::edit_distance::EditType *>(enum testing::internal::edit_distance::EditType * const,enum testing::internal::edit_distance::EditType * const,struct std::integral_constant<unsigned __int64,4>)" (??$_Reverse_unchecked1@PEAW4EditType@edit_distance@internal@testing@@@std@@YAXQEAW4EditType@edit_distance@internal@testing@@0U?$integral_constant@_K$03@0@@Z)
test_xtensor_blas.exe : fatal error LNK1120: 1 unresolved externals
NMAKE : fatal error U1077: 'C:\xtensor-conda\Library\bin\cmake.exe' : return code '0xffffffff'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\amd64\nmake.exe"' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\amd64\nmake.exe"' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\amd64\nmake.exe"' : return code '0x2'
```

**Comment by emmenlau** [comment]
Ah but the build error on Azure is related. I think my change only works with cmake >= 3.18. I'll investigate...

**Comment by emmenlau** [comment]
Ok, it seems my changes trigger a very rare problem with Visual Studio, that symbol `__std_reverse_trivially_swappable_4` is not resolved. Its no well known and I could not find a clear solution, but reports indicate that a linking is performed with a Visual Studio linker that is older than the one that produced the Google test binaries.

I'll try to build with a downloaded GTest to confirm this.

**Comment by JohanMabille** [comment]
> Ok, it seems my changes trigger a very rare problem with Visual Studio, that symbol __std_reverse_trivially_swappable_4 is not resolved. Its no well known and I could not find a clear solution, but reports indicate that a linking is performed with a Visual Studio linker that is older than the one that produced the Google test binaries.

Thanks for catching this, looks like this is the reason for the failure of https://github.com/xtensor-stack/xtensor-blas/pull/172

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -25,34 +27,6 @@ else()
...
-    _cxx_std_flag CXX REQUIRED
-    "-std=c++14"  # this should work with GNU, Intel, PGI
-    "/std:c++14"  # this should work with MSVC
-  )
-  message(STATUS "Building with C++14")
-endif()
-
-if(NOT _cxx_std_flag)
-  message(FATAL_ERROR "xtensor-blas needs a C++14-compliant compiler.")
-endif()
```
Comment (JohanMabille) [inline test/CMakeLists.txt:55]: Why did you need to change this logic? I think it was added to explicitely support "old" versions of CMake, is this change required?

**Comment by JohanMabille** [comment]
Looks like there are two unrelated changes here, improving gtest support and changes relative to C++17 support. Although these changes are small, they have enough impact on the codebase so that it makes sense to split the PR.

**Comment by emmenlau** [comment]
Yes you are right - I'll split and update the comments.

File: test/CMakeLists.txt (hunk 1 above)
Comment (emmenlau) [reply test/CMakeLists.txt]: See now the separate PR https://github.com/xtensor-stack/xtensor-blas/pull/175

**Pull request merged**